server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -263,6 +263,7 @@ static Mapping createDynamicUpdate(MappingLookup mappingLookup,
         RootObjectMapper root;
         if (dynamicMappers.isEmpty() == false) {
             root = createDynamicUpdate(mappingLookup, dynamicMappers);
+            root.fixRedundantIncludes();
```

Comment:
is there a reason that we don't always do this when we create a root object mapper?

Reply:
It would be nice to do this on RootObjectMapper.Builder.build(), but unfortunately RootObjectMapper (and indeed ObjectMapper in general) is not immutable, so we need to have this extra method.

Reply:
arg, I was initially thinking that calling this in the DocumentMapper construction would fix it but it does not. But why can't build call this method before returning the root?

Reply:
> But why can't build call this method before returning the root?

It could, but because the resulting object is not immutable it wouldn't help us here.  Mappers can still be added to the RootObjectMapper *after* build() has been called, and in fact that's precisely what happens in `createDynamicUpdate()`.

Reply:
I guess that makes sense now :)

Reply:
I heard somebody say we should use builders all the way, not sure who it was :)

Reply:
Long term we should change dynamic mappers to use builders and not mapper objects, but that's a different discussion :)